src/sync/mod.rs

```diff
@@ -0,0 +1,355 @@
+                .map_err(AnchoringChainUpdateError::Client)?;
+        }
+        Ok(())
+    }
+
+    fn find_private_key(
+        &self,
+        anchoring_keys: impl IntoIterator<Item = btc::PublicKey>,
+    ) -> Option<(btc::PublicKey, btc::PrivateKey)> {
+        anchoring_keys.into_iter().find_map(|public_key| {
+            self.key_pool
+                .get(&public_key)
+                .cloned()
+                .map(|private_key| (public_key, private_key))
+        })
+    }
+}
+
+/// Errors that occur when updating the sync with Bitcoin task.
+#[derive(Debug)]
+pub enum SyncWithBitcoinError<C: Display, R: Display> {
+    /// Error occurred in the private API client.
+    Client(C),
+    /// Error occurred in the Bitcoin relay.
+    Relay(R),
+    /// Internal error.
+    Internal(failure::Error),
+    /// Initial funding transaction is unconfirmed.
+    UnconfirmedFundingTransaction(Hash),
+}
+
+/// Pushes anchoring transactions to the Bitcoin blockchain.
+#[derive(Debug)]
+pub struct SyncWithBitcoinTask<T, R>
+where
+    T: PrivateApi + 'static,
+    R: BitcoinRelay + 'static,
+{
+    btc_relay: R,
+    api_client: T,
+}
+
+impl<T, R> SyncWithBitcoinTask<T, R>
+where
+    T: PrivateApi + 'static,
+    R: BitcoinRelay + 'static,
+    T::Error: Display,
+    R::Error: Display,
+{
+    /// Create a new sync with Bitcoin task instance.
+    pub fn new(btc_relay: R, api_client: T) -> Self {
+        Self {
+            api_client,
+            btc_relay,
+        }
+    }
+
+    /// Perform a one attempt to send the first uncommitted anchoring transaction into the Bitcoin network, if any.
+    /// sign an anchoring proposal, if any. Return an index of the first committed transaction.
+    pub fn process(
+        &self,
+        latest_committed_tx_index: Option<u64>,
+    ) -> Result<Option<u64>, SyncWithBitcoinError<T::Error, R::Error>> {
+        log::trace!("Perform syncing with the Bitcoin network");
+
+        let (index, tx) = {
+            if let Some(index) = latest_committed_tx_index {
+                // Check that the latest committed transaction was really committed into
+                // the Bitcoin network.
+                let tx = self.get_transaction(index)?;
+                if self.transaction_is_committed(tx.id())? {
+                    let chain_len = self
+                        .api_client
+                        .transactions_count()
+                        .map_err(SyncWithBitcoinError::Client)?
+                        .value;
+
+                    if index + 1 == chain_len {
+                        return Ok(Some(index));
+                    }
+                    let index = index + 1;
+                    (index, self.get_transaction(index)?)
+                } else {
+                    (index, tx)
+                }
+            }
+            // Perform to find the actual uncommitted transaction.
+            else if let Some((tx, index)) = self.find_index_of_first_uncommitted_transaction()? {
```

Comment:
Shouldn't this method be called `find_first_uncommitted_transaction` since it returns a transaction as well? Also, `(index, tx)` order of returned values is more expected, IMO.